### PR TITLE
drivers/dose: introduce watchdog timer

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Setting a timer for each received byte puts a low upper bound on the data rates that can be achieved with dose, as `xtimer_set` (or any timer set with can current timer API) be pretty expensive.
This prompted me to propose a new API for countdowns that are cheaper to set (#17159).

But we can do even better and *don't set any timer* inside the RX interrupt if we use a periodic timer instead.

In DOSE the timer on RX is used to abort the frame if no more bytes are received. Instead of setting a timer for each byte, we can instead use a periodic watchdog timer that checks if the RX buffer has received any new bytes since the last time the watchdog timer was executed.

This gives a good enough stall detection. If we have multiple DOSE interfaces we want to avoid too many interrupts, so let the DOSE watchdog timer check all DOSE interfaces. Since they are usually in an array, this is reasonably cheap.

### Testing procedure

With this I can use baud rates of 2 MBit/s between `samr21-xpro` and `same54-xpro`.
This is even if we speed up the TX path with hardware collision detection (#16681).

### CODE REVIEW

**Comment by kfessel** [comment]
Could this use higher level periodic timers like ztimer_periodic?

**Comment by benpicco** [comment]
> Could this use higher level periodic timers like ztimer_periodic?

This would require a dedicated thread which I'd rather avoid.
Also ztimer does not use a periodic HW timer but will instead always set a new timeout (which involved reading the current time) and is comparably expensive.

Most MCUs have plenty HW timers, so why not use them?

**Comment by kfessel** [comment]
please use the HW timer, i was just talking about the non watchdog-fallback (xtimer) that is still in dose, this could also profit from using the same priciple of not setting the timer every byte, but run a periodic timer (sys/include/ztimer/periodic.h)

**Comment by benpicco** [comment]
Ah nice I didn't know we have a `ztimer_periodic_init()` now.
But note that under the hood it will still [call `ztimer_set()`](https://github.com/RIOT-OS/RIOT/blob/master/sys/ztimer/periodic.c#L40) with each cycle, so it's just a convenience function, it does not save anything.

We might get rid of the xtimer calls for each byte on the send path (but then we will have to reduce the timer interval), but the random backoff can't be done without re-setting the timer, so I left the use of xtimer in place there.

**Comment by benpicco** [comment]
But it would allow us to use it as a fall-back if no `DOSE_TIMER_DEV` is defined. That way we can always use the watchdog functionality (instead of byte-wise timers) and  reduce code branches - @jue89 what do you think?

**Comment by benpicco** [comment]
I found out that this also requires 1d17f68d353c360dcdbd3656b53f09ad989053aa for proper collision avoidance - but I didn't backport that yet.

File: drivers/dose/Makefile.dep [hunk 1]
```diff
@@ -2,6 +2,10 @@ FEATURES_REQUIRED += periph_gpio_irq
 FEATURES_REQUIRED += periph_uart
 FEATURES_OPTIONAL += periph_uart_rxstart_irq
 
+ifneq (,$(filter dose_watchdog,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_timer_periodic
+endif
```
Comment (fjmolinas) [inline drivers/dose/Makefile.dep:7]: Can this get rid of the xtimer dependency?

File: drivers/dose/dose.c [hunk 2]
```diff
@@ -106,11 +106,100 @@ static inline void _disable_sense(dose_t *ctx)
 #endif
 }
 
+
```
Comment (fjmolinas) [inline drivers/dose/dose.c:109]:
```suggestion
```

File: drivers/dose/dose.c [hunk 3]
```diff
@@ -669,4 +771,14 @@ void dose_setup(dose_t *ctx, const dose_params_t *params, uint8_t index)
     DEBUG("dose timeout set to %" PRIu32 " µs\n", ctx->timeout_base);
     ctx->timeout.callback = _isr_xtimer;
     ctx->timeout.arg = ctx;
+
+#ifdef MODULE_DOSE_WATCHDOG
```
Comment (fjmolinas) [inline drivers/dose/dose.c:775]: Can you use `IS_USED`?

File: drivers/dose/dose.c [hunk 4]
```diff
@@ -106,11 +106,100 @@ static inline void _disable_sense(dose_t *ctx)
...
+static unsigned _wdt_period;
+#endif
+
+static inline void _watchdog_start(void)
+{
+    if (_watchdog_users) {
+        return;
+    }
+
+    ++_watchdog_users;
```
Comment (fjmolinas) [inline drivers/dose/dose.c:126]:
```suggestion
    _watchdog_users++;
```

Comment (fjmolinas) [reply drivers/dose/dose.c:126]: Feel free to ignore, I just find it weird to use when not needed, but just personal nitpicks.

**Review by fjmolinas — COMMENTED** [review]
Can you post some test-output and the test procedure?

> I found out that this also requires 1d17f68d353c360dcdbd3656b53f09ad989053aa for proper collision avoidance - but I didn't backport that yet. 

Can you explain a bit more? Should this PR includ that?

Otherwise LGTM

File: drivers/dose/Makefile.dep (hunk 1 above)
Comment (benpicco) [reply drivers/dose/Makefile.dep:8]: No xtimer is still used for the random backoff and as fallback if no HW timer is configured

File: drivers/dose/dose.c (hunk 3 above)
Comment (benpicco) [reply drivers/dose/dose.c:824]: Well the driver doesn't use `IS_USED` yet so I found `#ifdef` to be more consistent.

**Comment by benpicco** [comment]
Ok so now that all PRs are rebased I can test them individually.
This actually reveals a performance degradation with slow baud rates :confused: 

My simple test was just `ping -c 100 -i 25 ff02::1` with three boards on the bus:

```
# 115200 baud

## master

100 packets transmitted, 65 packets received, 9 duplicates, 35% packet loss
100 packets transmitted, 68 packets received, 9 duplicates, 32% packet loss
100 packets transmitted, 63 packets received, 11 duplicates, 37% packet loss

## master, CONFIG_DOSE_TIMEOUT_BYTES=16

100 packets transmitted, 99 packets received, 92 duplicates, 1% packet loss
100 packets transmitted, 99 packets received, 96 duplicates, 1% packet loss
100 packets transmitted, 100 packets received, 91 duplicates, 0% packet loss

## drivers/dose-watchdog HW watchdog

100 packets transmitted, 52 packets received, 13 duplicates, 48% packet loss
100 packets transmitted, 54 packets received, 9 duplicates, 46% packet loss
100 packets transmitted, 50 packets received, 9 duplicates, 50% packet loss

## drivers/dose-watchdog HW watchdog, CONFIG_DOSE_TIMEOUT_BYTES=16

100 packets transmitted, 98 packets received, 84 duplicates, 2% packet loss
100 packets transmitted, 99 packets received, 83 duplicates, 1% packet loss
100 packets transmitted, 99 packets received, 88 duplicates, 1% packet loss

## drivers/dose-watchdog, watchdog disabled

100 packets transmitted, 54 packets received, 10 duplicates, 46% packet loss
100 packets transmitted, 48 packets received, 6 duplicates, 52% packet loss
100 packets transmitted, 46 packets received, 9 duplicates, 54% packet loss

## drivers/dose-watchdog, watchdog disabled, CONFIG_DOSE_TIMEOUT_BYTES=16

100 packets transmitted, 90 packets received, 64 duplicates, 10% packet loss
100 packets transmitted, 0 packets received, 100% packet loss
100 packets transmitted, 0 packets received, 100% packet loss

## drivers/dose-collision+watchdog+ringbuffer

100 packets transmitted, 38 packets received, 13 duplicates, 62% packet loss
100 packets transmitted, 41 packets received, 15 duplicates, 59% packet loss
100 packets transmitted, 45 packets received, 11 duplicates, 55% packet loss

## drivers/dose-collision+watchdog+ringbuffer, CONFIG_DOSE_TIMEOUT_BYTES=16

100 packets transmitted, 100 packets received, 89 duplicates, 0% packet loss
100 packets transmitted, 100 packets received, 81 duplicates, 0% packet loss
100 packets transmitted, 100 packets received, 86 duplicates, 0% packet loss
```

will investigate.

**Comment by benpicco** [comment]
Ok so with the latest changes I get

```
## drivers/dose-watchdog, HW watchdog

100 packets transmitted, 92 packets received, 35 duplicates, 8% packet loss
100 packets transmitted, 87 packets received, 38 duplicates, 13% packet loss
100 packets transmitted, 91 packets received, 29 duplicates, 9% packet loss

## drivers/dose-watchdog, xtimer watchdog

100 packets transmitted, 93 packets received, 35 duplicates, 7% packet loss
100 packets transmitted, 88 packets received, 31 duplicates, 12% packet loss
100 packets transmitted, 89 packets received, 33 duplicates, 11% packet loss

## drivers/dose-watchdog, watchdog disabled - 448 bytes of ROM less

100 packets transmitted, 84 packets received, 26 duplicates, 16% packet loss
100 packets transmitted, 96 packets received, 38 duplicates, 4% packet loss
100 packets transmitted, 91 packets received, 34 duplicates, 9% packet loss
```

which *is* an improvement compared to `master`.

File: drivers/dose/dose.c [hunk 5]
```diff
@@ -107,11 +111,112 @@ static inline void _disable_sense(dose_t *ctx)
...
+{
+    if (_watchdog_users) {
+        return;
+    }
+
+    _watchdog_users++;
+#ifdef DOSE_TIMER_DEV
+    timer_start(DOSE_TIMER_DEV);
+#else
+    xtimer_set(&_wdt, _wdt_period);
```
Comment (benpicco) [inline drivers/dose/dose.c:134]: The software timer fallback doesn't improve anything compared to not using `dose_watchdog`, but we could use it to eliminate the need for a `dose_watchdog` pseudo-module and always enable the watchdog code.

Comment (fjmolinas) [reply drivers/dose/dose.c:134]: If it doesn't improve anything could we remove it?

**Review by fjmolinas — COMMENTED** [review]
Looks good, I'll trust your testing. But if you say the xtimer backup does not help then why not keep it? It's just adding rom no?

**Comment by benpicco** [comment]
Sure. We could also `s/MODULE_DOSE_WATCHDOG/DOSE_TIMER_DEV/` to get rid of the `dose_watchdog` pseudo-module, but then we would always enable `periph_timer_periodic` even if no watchdog timer is defined.

File: drivers/dose/dose.c [hunk 6]
```diff
@@ -107,11 +111,88 @@ static inline void _disable_sense(dose_t *ctx)
...
+#endif
+
 static dose_signal_t state_transit_blocked(dose_t *ctx, dose_signal_t signal)
 {
+    (void) signal;
     uint32_t backoff;
+
+    backoff = random_uint32_range(xtimer_usec_from_ticks(xtimer_min_timeout),
+                                  2 * ctx->timeout_base);
+    xtimer_set(&ctx->timeout, backoff);
```
Comment (fjmolinas) [inline drivers/dose/dose.c:184]: I see that the backoff scheme has been changed, before there was a bias towards receiving nodes, can you explain this change a bit?

Comment (fjmolinas) [reply drivers/dose/dose.c:184]: Thanks!

**Comment by fjmolinas** [comment]
Please squash @benpicco!

**Comment by fjmolinas** [comment]
> Please squash @benpicco!

And trigger the ci!

**Review by fjmolinas — APPROVED** [review]
ACK! @benpicco do you mind doing a final test?

**Comment by benpicco** [comment]
Sure, test between `same54-xpro`, `adafruit-itsybitsy-m4` and `nrf52840dk` at default baud rates:

```
2022-01-18 22:16:29,235 # ping ff02::1
2022-01-18 22:16:29,262 # 12 bytes from fe80::a884:b7ff:fea4:e6a1%5: icmp_seq=0 ttl=64 time=19.791 ms
2022-01-18 22:16:29,305 # 12 bytes from fe80::ac0b:2bff:feab:bd96%5: icmp_seq=0 ttl=64 time=62.696 ms (DUP!)
2022-01-18 22:16:30,268 # 12 bytes from fe80::a884:b7ff:fea4:e6a1%5: icmp_seq=1 ttl=64 time=25.921 ms
2022-01-18 22:16:30,283 # 12 bytes from fe80::ac0b:2bff:feab:bd96%5: icmp_seq=1 ttl=64 time=40.793 ms (DUP!)
2022-01-18 22:16:31,269 # 12 bytes from fe80::ac0b:2bff:feab:bd96%5: icmp_seq=2 ttl=64 time=26.724 ms
2022-01-18 22:16:31,269 # 
2022-01-18 22:16:31,272 # --- ff02::1 PING statistics ---
2022-01-18 22:16:31,278 # 3 packets transmitted, 3 packets received, 2 duplicates, 0% packet loss
2022-01-18 22:16:31,282 # round-trip min/avg/max = 19.791/35.185/62.696 ms

```

**Comment by fjmolinas** [comment]
Goooo!

**Pull request merged**